private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -434,4 +480,171 @@ public boolean hasRequestMetadataOnly() {
+          );
+        } else {
+          LOG.info(String.format("PGP signature verified for file %s", toSign.getFileName()));
+        }
+      } catch (PGPException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    return signatureOutputPath;
+  }
+
+  public static String bytesToHex(byte[] bytes) {
+    return String.format("%0" + (bytes.length << 1) + "x", new BigInteger(1, bytes));
+  }
+
+  public static boolean verify(Path originFilePath, PGPSignature signature, PGPPublicKey publicKey) throws PGPException {
+    signature.init(new BcPGPContentVerifierBuilderProvider(), publicKey);
+
+    LOG.info(String.format("[Verify] Public Key Fingerprint: %s", bytesToHex(publicKey.getFingerprint())));
+
+    byte[] buffer = new byte[BUFFER];
+    int len;
+    try (InputStream in = new BufferedInputStream(Files.newInputStream(originFilePath))) {
+      while ((len = in.read(buffer)) >= 0) {
+        signature.update(buffer, 0, len);
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+    return signature.verify();
+  }
+
+  public static PGPSignatureList readSignatures(Path signaturePath) {
+    try (
+            InputStream stream = new BufferedInputStream(Files.newInputStream(signaturePath));
+            InputStream decoderStream = PGPUtil.getDecoderStream(stream)
+    ) {
+      return readSignatureList(decoderStream);
+    } catch (IOException | PGPException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static PGPSignatureList readSignatureList(InputStream decoderStream) throws IOException, PGPException {
+    PGPObjectFactory objectFactory = new PGPObjectFactory(decoderStream, new BcKeyFingerprintCalculator());
+    Object nextObject = objectFactory.nextObject();
+    if (nextObject instanceof PGPSignatureList) {
+      return (PGPSignatureList) nextObject;
+    } else if (nextObject instanceof PGPCompressedData) {
+      return readSignatureList(((PGPCompressedData) nextObject).getDataStream());
+    } else {
+      LOG.warning(String.format("Skip parsing signature list in %s.", nextObject == null ? "invalid file" : nextObject.getClass()));
+      return null;
+    }
+  }
+
+  private static PGPPrivateKey createPrivateKey(PGPSecretKey secretKey, String password) {
+    try {
+      PBESecretKeyDecryptor decryptor = new BcPBESecretKeyDecryptorBuilder(new BcPGPDigestCalculatorProvider()).build(password.toCharArray());
+      return secretKey.extractPrivateKey(decryptor);
+    } catch (PGPException e) {
+      throw new RuntimeException("Could not extract private key", e);
+    }
+  }
+
+  private enum SigningMethod {
+    GPG,
+    PGP,
+    NONE
+  }
+
+  private static class SigningMetadata {
+    private final String signingKey;
+    private final String signingPassword;
+    private final SigningMethod signingMethod;
+
+    public SigningMetadata(Boolean gpgSign, Boolean useInMemoryPgpKeys, String signingKey, String signingPassword) {
```

Comment:
Are you referring to `SigningMetadata` class? I created it mainly to avoid extending the param list of this upload method and other similar methods. 
```
              upload(
                  repo,
                  credentials,
                  coords,
                  String.format("-%s.%s", classifier, ext),
                  artifact,
                  signingMetadata));
```

Reply:
I mean the use of `Boolean` instead of `boolean`. We never handle the `null` case, so not using the primitive type means that we could get a `NullPointerException` if we unbox `useInMemoryPgpKeys` and `gpgSign` in this method. Using the primitive type prevents this.

Reply:
Nice catch, updated